File: dotnet/src/ProcessEngine.ExternalTaskAPI.Client/ExternalTaskApiClientService.cs
namespace ProcessEngine.ExternalTaskAPI.Client
{
    using System;
    using System.Collections.Generic;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Text;
    using System.Threading.Tasks;

    using Foundation.IAM.Contracts;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Serialization;
    using ProcessEngine.ExternalTaskAPI.Client.Requests;
    using ProcessEngine.ExternalTaskAPI.Contracts;

    public class ExternalTaskApiClientService : IExternalTaskAPI
    {
        private const string BaseRoute = "api/external_task/v1";

        private readonly HttpClient httpClient;


        public ExternalTaskApiClientService(HttpClient httpClient)
        {
            this.httpClient = httpClient;
        }

        public async Task ExtendLock(IIdentity identity, string workerId, string externalTaskId, int additionalDuration)
        {
            string uri = $"task/{externalTaskId}/extend_lock";
            ExtendLockRequest request = new ExtendLockRequest
            (
                workerId,
                additionalDuration
            );

            await this.SendPostToExternalTaskApi(identity, uri, request);
        }

        public async Task<IEnumerable<ExternalTask<TPayload>>> FetchAndLockExternalTasks<TPayload>(IIdentity identity, string workerId, string topicName, int maxTasks, int longPollingTimeout, int lockDuration) where TPayload : new()
        {
            string uri = "fetch_and_lock";
            FetchAndLockRequest request = new FetchAndLockRequest
            (
                workerId,
                topicName,
                maxTasks,
                longPollingTimeout,
                lockDuration
            );

            IEnumerable<ExternalTask<TPayload>> response = await this.SendPostToExternalTaskApi<FetchAndLockRequest, IEnumerable<ExternalTask<TPayload>>>
            (
                identity,
                uri,
                request
            );
            return response;
        }

        public async Task FinishExternalTask<TPayload>(IIdentity identity, string workerId, string externalTaskId, TPayload payload)
        {
            string uri = $"task/{externalTaskId}/finish";

            FinishExternalTaskRequest<TPayload> request = new FinishExternalTaskRequest<TPayload>
            (
                workerId,
                payload
            );

            await this.SendPostToExternalTaskApi(identity, uri, request);
        }

        public async Task HandleBpmnError(IIdentity identity, string workerId, string externalTaskId, string errorCode)
        {
            string uri = $"task/{externalTaskId}/handle_bpmn_error";

            HandleBpmnErrorRequest request = new HandleBpmnErrorRequest
            (
                workerId,
                errorCode
            );

            await this.SendPostToExternalTaskApi(identity, uri, request);
        }

        public async Task HandleServiceError(IIdentity identity, string workerId, string externalTaskId, string errorMessage, string errorDetails)
        {
            string uri = $"task/{externalTaskId}/handle_service_error";

            HandleServiceErrorRequest request = new HandleServiceErrorRequest
            (
                workerId,
                errorMessage,
                errorDetails
            );

            await this.SendPostToExternalTaskApi(identity, uri, request);
        }

        public void Dispose()
        {
            this.httpClient.Dispose();
        }

        private async Task<TResponse> SendPostToExternalTaskApi<TRequest, TResponse>(IIdentity identity, string uri, TRequest request)
        {
            SetAuthenticationHeader(identity);

            StringContent content = GetRequestAsStringContent(request);

            var response = await this.httpClient.PostAsync(CombineBaseRouteWith(uri), content);

            response.EnsureSuccessStatusCode();

            return await DeserializeResposne<TResponse>(response);
        }

        private async Task SendPostToExternalTaskApi<TRequest>(IIdentity identity, string uri, TRequest request)
        {
            SetAuthenticationHeader(identity);

            StringContent content = GetRequestAsStringContent(request);

            var response = await this.httpClient.PostAsync(CombineBaseRouteWith(uri), content);

            response.EnsureSuccessStatusCode();
        }

        private void SetAuthenticationHeader(IIdentity identity)
        {
            this.httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", identity.ToString());
        }

        private StringContent GetRequestAsStringContent<TRequest>(TRequest request)
        {
            var settings = new JsonSerializerSettings
            {
                ContractResolver = new CamelCasePropertyNamesContractResolver()
            };

            string serializedRequest = JsonConvert.SerializeObject(request, settings);
            StringContent content = new StringContent(serializedRequest, Encoding.UTF8, "application/json");
            return content;
        }

        private async Task<TResponse> DeserializeResposne<TResponse>(HttpResponseMessage response)
        {
            string serializedResponse = await response.Content.ReadAsStringAsync();

            return JsonConvert.DeserializeObject<TResponse>(serializedResponse);
        }

        private string CombineBaseRouteWith(string uri)
        {
            return $"{BaseRoute}/{uri}";
        }
    }
}

File: python/external_task_api_client/external_task_api_client_service.py
import aiohttp
import asyncio
import json


class ExternalTaskApiClientService:

    def __init__(self, baseUrl):
        self.__baseUrl = baseUrl

    async def extendLock(self, identity, workerId, externalTaskId, additionalDuration):
        uri = "task/" + externalTaskId + "/extend_lock"
        request = {
            "workerId": workerId,
            "additionalDuration": additionalDuration
        }

        return await self.__sendPostToExternalTaskApi(identity, uri, request)

    async def fetchAndLockExternalTasks(self, identity, workerId, topicName, maxTasks, longPollingTimeout, lockDuration):
        uri = "fetch_and_lock"
        request = {
            "workerId": workerId,
            "topicName": topicName,
            "maxTasks": maxTasks,
            "longPollingTimeout": longPollingTimeout,
            "lockDuration": lockDuration
        }

        return await self.__sendPostToExternalTaskApi(identity, uri, request)

    async def finishExternalTask(self, identity, workerId, externalTaskId, payload):
        uri = "task/" + externalTaskId + "/finish"
        request = {
            "workerId": workerId,
            "result": payload
        }

        await self.__sendPostToExternalTaskApi(identity, uri, request)

    async def handleBpmnError(self, identity, workerId, externalTaskId, errorCode):
        uri = "task/" + externalTaskId + "/handle_bpmn_error"
        request = {
            "workerId": workerId,
            "errorCode": errorCode
        }

        await self.__sendPostToExternalTaskApi(identity, uri, request)

    async def handleServiceError(self, identity, workerId, externalTaskId, errorMessage, errorDetails):
        uri = "task/" + externalTaskId + "/handle_service_error"
        request = {
            "workerId": workerId,
            "errorMessage": errorMessage,
            "errorDetails": errorDetails
        }

        await self.__sendPostToExternalTaskApi(identity, uri, request)

    async def __sendPostToExternalTaskApi(self, identity, uri, request):
        headers = {"Authorization": "Bearer " + identity["token"]}
        url = self.__combineWithBaseUrl(uri)

        async with aiohttp.ClientSession(headers=headers) as session:
            async with session.post(url, json=request) as response:
                response.raise_for_status()
                if response.status == 200:
                    return await response.json()

    def __combineWithBaseUrl(self, uri):
        return self.__baseUrl + "/api/external_task/v1/" + uri
